libdrgn/elf_symtab.c

```diff
@@ -46,6 +58,98 @@ static struct drgn_error *find_elf_file_symtab(struct drgn_elf_file *file,
 	return NULL;
 }
 
+#ifdef WITH_LZMA
+static struct drgn_error *
+drgn_error_lzma(lzma_ret code)
+{
+	switch (code) {
+		case LZMA_MEM_ERROR:
+			return &drgn_enomem;
+		case LZMA_OPTIONS_ERROR:
+			return drgn_error_format(DRGN_ERROR_INVALID_ARGUMENT,
+						 "lzma: invalid options");
+		case LZMA_FORMAT_ERROR:
+		case LZMA_DATA_ERROR:
+		case LZMA_BUF_ERROR:
+			return drgn_error_format(DRGN_ERROR_INVALID_ARGUMENT,
+						 "lzma: format error (%d)", code);
+		default:
+			return drgn_error_format(DRGN_ERROR_OTHER,
+						 "lzma: unknown error (%d)", code);
+	}
+}
+
+static struct drgn_error *
+load_gnu_debugdata_file(struct drgn_module *module, Elf_Scn *gnu_debugdata_scn)
+{
+	if (module->gnu_debugdata_file)
+		return NULL;
+
+	Elf_Data *gnu_debugdata_data;
+	struct drgn_error *err;
+	err = read_elf_section(gnu_debugdata_scn, &gnu_debugdata_data);
+	if (err)
+		return err;
+
+	_cleanup_(lzma_end) lzma_stream stream = LZMA_STREAM_INIT;
+	lzma_ret ret = lzma_stream_decoder(&stream, UINT64_MAX, 0);
+	if (ret != LZMA_OK)
+		return drgn_error_lzma(ret);
+
+	stream.next_in = gnu_debugdata_data->d_buf;
+	stream.avail_in = gnu_debugdata_data->d_size;
+
+	size_t datasize = 2 * gnu_debugdata_data->d_size;
+	_cleanup_free_ void *data = malloc(datasize);
+
+	stream.next_out = data;
+	stream.avail_out = gnu_debugdata_data->d_size;
+
+	size_t bytes_decoded;
+	while (1) {
+		ret = lzma_code(&stream, LZMA_RUN);
+		if (ret != LZMA_OK && ret != LZMA_STREAM_END)
+			return drgn_error_lzma(ret);
+
+		bytes_decoded = (size_t)stream.next_out - (size_t)data;
+		if (ret == LZMA_STREAM_END)
+			datasize = bytes_decoded;
+		else
+			datasize *= 2;
+
+		data = realloc(data, datasize);
+		if (!data)
+			return &drgn_enomem;
+
+		stream.next_out = (uint8_t *)data + bytes_decoded;
+		stream.avail_out = datasize - bytes_decoded;
+		if (ret == LZMA_STREAM_END)
+			break;
```

Comment:
A few things here:

* If the `realloc()` fails, you're leaking `data`.
* If the `realloc()` fails in the `LZMA_STREAM_END` case, it doesn't need to be fatal.
* Should again be paranoid about `datasize` overflow.

Maybe something like:

```suggestion
		if (ret == LZMA_STREAM_END) {
			// Try shrinking to fit.
			void *tmp = realloc(data, bytes_decoded);
			if (tmp)
				data = tmp;
			break;
		} else if (__builtin_mul_overflow(datasize, 2U, &datasize)) {
			return &drgn_enomem;
		} else {
			void *tmp = realloc(data, datasize);
			if (!tmp)
				return &drgn_enomem;
			data = tmp;
			stream.next_out = (uint8_t *)data + bytes_decoded;
			stream.avail_out = datasize - bytes_decoded;
		}
```